### PR TITLE
DependsOnMethods not respected on the sequential run on 2nd test with same base testClass

### LINKED ISSUE #1648: Depends on method is not respected on the sequential run on second test that extends same base testClass
Maven surefire  2.20.1 2.20
I have Checked only those versions 

I have two concrete Test Classes 
TestOne
TestTwo
They both extend class B that extends Class A. 
Abstract class A and B have 2 test each and I need to run them in order: 
A -> Test one then Test two
 then 
      B - Test one then test two

If I run the test using IDE one by one all is fine test run in order same if I run using maven.

If I run those test together[single thread one at the time] in IDE or maven The first Test run tests in order but the second Test the order is not the same and as they extend same Base class should be the same.

If I run the test in parallel the order is preserved..

Could you kindly advice how to make sure when running the test sequentially the order of execution is preserved even after the first test running [that extends the same base]

### Expected behavior
```
A TestOne 1 ........
A test Two  ........
B test 1 ........
B test 2........

A TestOne 1 ........
A test Two  ........
B test 1 ........
B test 2........
```

### Actual behavior
"B test 1" is executed after "A TestOne 1" and should be executed after "A test Two "

### Is the issue reproductible on runner?

- [x] Maven
- [x] IntelliJ

```
A TestOne 1 ........
A test Two  ........
B test 1 ........
B test 2........
```
-----------------------------
```
A TestOne 1 ........
B test 1 ........
A test Two  ........
B test 2........
```

Note:
Maven Profile
threadCount = 1
maven surefire plugin 
```xml
                             <forkCount>${threadCount}</forkCount>
                            <parallel>suites</parallel>
                            <threadCountClasses>${threadCount}</threadCountClasses>
```

```java
public abstract class A {
    @Test
    protected void test1(){       System.out.println("A TestOne 1 ........");    }

    @Test(dependsOnMethods = {"test1"} )
    protected void test2(){        System.out.println("A test Two  ........");    }
}

public abstract class B extends A {
    @Test(dependsOnMethods = {"test2"})
    protected void test3(){     System.out.println("B test 1 ........"); }

    @Test(dependsOnMethods = {"test3"})
    protected void test4(){  System.out.println("B test 2........");    }
}

public class TestOne extends B {
}
public class TestTwo extends B {
}
```

Thank you for your help

### PR DESCRIPTION
**Pull request by krmahadevan**
Closes #1648

Fixes #1648 .

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

### CODE REVIEW

File: src/main/java/org/testng/DependencyMap.java [hunk 1]
```diff
@@ -61,12 +61,15 @@ public ITestNGMethod getMethodDependingOn(String methodName, ITestNGMethod fromM
...
+              fromMethodInstance.getClass().isAssignableFrom(mInstance.getClass())) {
+          return m;
+      } else {
       // otherwise, it's a method depending on a method in a different class so we
       // don't bother checking the instance
-      if (fromMethod.getRealClass().isAssignableFrom(m.getRealClass())) {
-        if (m.getInstance() == fromMethod.getInstance()) return m;
-      } else {
-        return m;
+          return m;
```
Comment (juherr) [inline src/main/java/org/testng/DependencyMap.java]: indentation is only 2 spaces in this file.

File: src/main/java/org/testng/DependencyMap.java [hunk 2]
```diff
@@ -61,12 +61,15 @@ public ITestNGMethod getMethodDependingOn(String methodName, ITestNGMethod fromM
...
       // If they are in the same class hierarchy, they must belong to the same instance,
+      Object fromMethodInstance = fromMethod.getInstance();
+      Object mInstance = m.getInstance();
+      if (mInstance != null && fromMethodInstance != null &&
+              fromMethodInstance.getClass().isAssignableFrom(mInstance.getClass())) {
+          return m;
+      } else {
       // otherwise, it's a method depending on a method in a different class so we
       // don't bother checking the instance
-      if (fromMethod.getRealClass().isAssignableFrom(m.getRealClass())) {
```
Comment (juherr) [inline src/main/java/org/testng/DependencyMap.java]: What was the issue?
If I understand the change well, a minimal diff could be:

```java
if (fromMethod.getRealClass().isAssignableFrom(m.getRealClass()) &&
  m.getInstance() != null && fromMethod.getInstance() !== null) {
  return m;
}
```

**Review by juherr — COMMENTED** [review]
Why did it work well with parallel and not sequential? Any idea?

File: src/main/java/org/testng/DependencyMap.java (hunk 2 above)
Comment (krmahadevan) [reply src/main/java/org/testng/DependencyMap.java]: @juherr - The fix was to consider the instance class and not the declaring class which is what `getRealClass()` returns. In cases of inheritance (as in this case), we should have been checking if the method upon which a dependent method was executed for the current instance. TestNG was not doing that. That is what has been changed. See [here](https://stackoverflow.com/a/10028333) for some examples when involving inheritance and `getDeclaringClass`

File: src/main/java/org/testng/DependencyMap.java (hunk 1 above)
Comment (krmahadevan) [reply src/main/java/org/testng/DependencyMap.java]: Yes I know. But its hard to follow that indentation especially in other places the IDE is configured to 4 spaces. So I chose to leave this as is. I think on a long term basis, we need to get this sorted out across all classes which brings us back to our earlier discussion on introducing a IDE formatter file for both IntelliJ and eclipse.

**Comment by krmahadevan** [comment]
@juherr - I have responded to both your observations. Please let me know if that makes sense.

**Comment by krmahadevan** [comment]
ping @juherr - Can you please advise on this. I have rebased off of master to resolve conflicts and pushed in the amended commits.

File: src/main/java/org/testng/DependencyMap.java [hunk 3]
```diff
@@ -61,11 +61,14 @@ public ITestNGMethod getMethodDependingOn(String methodName, ITestNGMethod fromM
...
       // If they are in the same class hierarchy, they must belong to the same instance,
-      // otherwise, it's a method depending on a method in a different class so we
-      // don't bother checking the instance
-      if (fromMethod.getRealClass().isAssignableFrom(m.getRealClass())) {
-        if (m.getInstance() == fromMethod.getInstance()) return m;
+      Object fromMethodInstance = fromMethod.getInstance();
+      Object mInstance = m.getInstance();
+      if (mInstance != null && fromMethodInstance != null &&
+              fromMethodInstance.getClass().isAssignableFrom(mInstance.getClass())) {
+        return m;
```
Comment (juherr) [inline src/main/java/org/testng/DependencyMap.java]: Strange: it looks we are returning the same `m` in the if and else statement.

Comment (krmahadevan) [reply src/main/java/org/testng/DependencyMap.java]: Yes @juherr . Its kind of like the earlier logic. I have merely included the `Object` into the equation, which wasn't the case earlier. I haven't figured out an alternative for this yet.

Comment (juherr) [reply src/main/java/org/testng/DependencyMap.java]: If both if and else are doing the same thing, we can remove the if and directly return the value ?

Comment (krmahadevan) [reply src/main/java/org/testng/DependencyMap.java]: @juherr - That doesn't work. I refactored it to make it a bit more explicit. I have amended the changes and pushed it in again. Please have a look.

**Comment by krmahadevan** [comment]
Rebased off of master, resolved merge conflicts and pushed in the changes. @juherr please help take a look

File: src/main/java/org/testng/DependencyMap.java [hunk 4]
```diff
@@ -60,19 +58,38 @@ public ITestNGMethod getMethodDependingOn(String methodName, ITestNGMethod fromM
     	return fromMethod;
     }
     for (ITestNGMethod m : l) {
-      // If they are in the same class hierarchy, they must belong to the same instance,
-      // otherwise, it's a method depending on a method in a different class so we
-      // don't bother checking the instance
-      if (fromMethod.getRealClass().isAssignableFrom(m.getRealClass())) {
-        if (m.getInstance() == fromMethod.getInstance()) return m;
-      } else {
+      if (isSameInstance(fromMethod, m)) {
```
Comment (juherr) [inline src/main/java/org/testng/DependencyMap.java]: Just merge the 2:
`if (isSameInstance(fromMethod, m) || belongToDifferentClassHierarchy(fromMethod, m) || hasNoInstance(fromMethod, m))`

Comment (krmahadevan) [reply src/main/java/org/testng/DependencyMap.java]: Yeah that looks more elegant and clean. Let me see if I can keep the tests happy with this change :)

Comment (krmahadevan) [reply src/main/java/org/testng/DependencyMap.java]: @juherr - I have changed it as per your recommendations. Please check

File: src/main/java/org/testng/DependencyMap.java [hunk 5]
```diff
@@ -63,16 +61,32 @@ public ITestNGMethod getMethodDependingOn(String methodName, ITestNGMethod fromM
...
+  private static boolean hasNoInstance(ITestNGMethod baseClassMethod, ITestNGMethod derivedClassMethod) {
+    Object baseInstance = baseClassMethod.getInstance();
+    Object derivedInstance = derivedClassMethod.getInstance();
+    return derivedInstance != null || baseInstance != null;
+  }
+
+  private static boolean isSameInstance(ITestNGMethod baseClassMethod, ITestNGMethod derivedClassMethod) {
+    Object baseInstance = baseClassMethod.getInstance();
+    Object derivedInstance = derivedClassMethod.getInstance();
+    return derivedInstance != null &&
```
Comment (juherr) [inline src/main/java/org/testng/DependencyMap.java]: Ok for me but I think we can merge `hasNoInstance` and `isSameInstance`. WDYT?

File: src/main/java/org/testng/DependencyMap.java [hunk 6]
```diff
@@ -63,16 +61,32 @@ public ITestNGMethod getMethodDependingOn(String methodName, ITestNGMethod fromM
...
   }
 
+  private static boolean belongToDifferentClassHierarchy(ITestNGMethod baseClassMethod, ITestNGMethod derivedClassMethod) {
+    return ! baseClassMethod.getRealClass().isAssignableFrom(derivedClassMethod.getRealClass());
+  }
+
+  private static boolean hasNoInstance(ITestNGMethod baseClassMethod, ITestNGMethod derivedClassMethod) {
+    Object baseInstance = baseClassMethod.getInstance();
+    Object derivedInstance = derivedClassMethod.getInstance();
+    return derivedInstance != null || baseInstance != null;
```
Comment (juherr) [inline src/main/java/org/testng/DependencyMap.java]: `return derivedInstance == null && baseInstance == null;` looks easier to understand according to the method name.

Comment (krmahadevan) [reply src/main/java/org/testng/DependencyMap.java]: @juherr - The logic cannot be changed, else it will cause test failures (especially the new one that was added and what this PR intends to fix). The method name actually meant that either `baseClassMethod` (or) `derivedClassMethod` *has no instance*

File: src/main/java/org/testng/DependencyMap.java (hunk 5 above)
Comment (krmahadevan) [reply src/main/java/org/testng/DependencyMap.java]: @juherr - I have already tried doing that. It breaks either the PR changes (or) breaks tests that are part of  `org.testng.internal.DynamicGraphHelperTest`

So we will have to stick to this.

File: src/main/java/org/testng/DependencyMap.java (hunk 6 above)
Comment (juherr) [reply src/main/java/org/testng/DependencyMap.java]: Ok, but `derivedInstance != null || baseInstance != null` means one of the 2 has an instance, right?

Comment (krmahadevan) [reply src/main/java/org/testng/DependencyMap.java]: Yes, that's one interpretation. The other would be, that one of them has no instance when `!=null` fails. Like I said, I think the method name is a bit confusing. Am open to any other suggestions

Comment (juherr) [reply src/main/java/org/testng/DependencyMap.java]: In fact, the name should be `hasInstance` because true is returned if `baseClassMethod` or `derivedClassMethod` have an instance.

Comment (krmahadevan) [reply src/main/java/org/testng/DependencyMap.java]: @juherr - I have renamed the method to `hasInstance`

**Review by juherr — APPROVED** [review]
